lib/server.js

```diff
@@ -232,10 +232,20 @@ class Server extends KarmaEventEmitter {
         let newBrowser = info.id ? (capturedBrowsers.getById(info.id) || singleRunBrowsers.getById(info.id)) : null
 
         if (newBrowser) {
+          // By default if a browser disconnects while still executing, we assume that the test
+          // execution still continues because just the socket connection has been terminated. Now
+          // since we know whether this is just a socket reconnect or full client reconnect, we
+          // need to update the browser state accordingly. This is necessary because in case a
+          // browser crashed and has been restarted, we need to start with a fresh execution.
+          if (!info.isSocketReconnect) {
+            newBrowser.setState(Browser.STATE_DISCONNECTED)
+          }
+
           newBrowser.reconnect(socket)
 
-          // We are restarting a previously disconnected browser.
-          if (newBrowser.state === Browser.STATE_DISCONNECTED && config.singleRun) {
```

Comment:
What it a bug that this was using `DISCONNECTED` before, or is this a change in behavior?

Reply:
Yeah that was a bug. The `reconnect` function always sets the state to `CONNECTED` if the state was `STATE_DISCONNECTED`. So it would never _pass_ this check.

See [the `reconnect` function source](https://github.com/karma-runner/karma/blob/b4831c190909960095e3e28234c1a207ccecb6be/lib/browser.js#L122-L145)